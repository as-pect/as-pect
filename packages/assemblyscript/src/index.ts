export type ASC = typeof import("assemblyscript/dist/asc.js") ;
export type AS = typeof import("assemblyscript/dist/assemblyscript.js");
export type TRANSFORM = typeof import("assemblyscript/dist/transform.js");

// @ts-ignore: Adding an aspect global
if (!globalThis.__aspect) globalThis.__aspect = {};

// We need to garuntee we use only a single version of the compiler

// @ts-ignore: Adding an aspect global for asc
if (!globalThis.__aspect.asc) globalThis.__aspect.asc = await import("assemblyscript/dist/asc.js");
// @ts-ignore: Adding an aspect global for asc
export const asc: ASC = globalThis.__aspect.asc;

// @ts-ignore: Adding an aspect global for asc
if (!globalThis.__aspect.assemblyscript) globalThis.__aspect.assemblyscript = await import("assemblyscript/dist/assemblyscript.js");
// @ts-ignore: Adding an aspect global for asc
export const assemblyscript: AS = globalThis.__aspect.assemblyscript;

// @ts-ignore: Adding an aspect global for asc
if (!globalThis.__aspect.transform) globalThis.__aspect.transform = await import("assemblyscript/dist/transform.js");
// @ts-ignore: Adding an aspect global for asc
export const transform: TRANSFORM = globalThis.__aspect.transform;

