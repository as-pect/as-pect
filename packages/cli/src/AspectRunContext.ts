import { IAspectConfiguration } from "./IAspectConfiguration";
import { IProcessConfiguration } from "./IProcessConfiguration";
import { IWritable } from "@as-pect/core/src/util/IWritable";

export interface IASCProps {
  stdout: IWritable;
  stderr: IWritable;
  listFiles(dirname: string, baseDir: string): string[];
  readFile(filename: string, baseDir: string): Uint8Array | null;
  writeFile(name: string, contents: Uint8Array, baseDir?: string): void;
}

export interface IASC {
  ready(): Promise<void>;
  main(
    args: string[],
    config: IASCProps,
    callback: (error: Error) => number,
  ): void;
}


export interface IOptionalRunContextProps {
  /** The assemblyscript help text generation function. */
  help: typeof import("assemblyscript/cli/util/options").help;
  /** The assemblyscript cli parse function. */
  parse: typeof import("assemblyscript/cli/util/options").parse;
  /** The result of calling the assemblyscript cli parse function. */
  parseResult: import("assemblyscript/cli/util/options").Result;
  /** The retreived as-pect config file. */
  config: Partial<IAspectConfiguration>;
  /** The assemblyscript compiler cli interface itself. */
  asc: IASC;
}

export interface IRequiredRunContextProps {
  /** Raw argv from the node process. */
  argv: string[];
  /** A set of file promises that must be resolved before the application exits. */
  files: Promise<void>[];
  /** A stripped down version of a process that contains an exit function, stdout and stderr. */
  process: IProcessConfiguration;
  /** The exit code used to exit the program. It starts as 0, and can only be set to 1 otherwise. */
  exitCode: number;
}

export type AspectRunContext = Partial<IOptionalRunContextProps> &
  IRequiredRunContextProps;
