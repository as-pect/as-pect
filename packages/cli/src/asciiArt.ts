import chalk from "chalk-template";

/**
 * @ignore
 *
 * This method prints the ascii art.
 * @param {string} version - The cli version
 */
export function printAsciiArt(): void {
  console.log(chalk`{bold.bgWhite.black ${""}       ___   _____                       __    
      /   | / ___/      ____  ___  _____/ /_   
     / /| | \\__ \\______/ __ \\/ _ \\/ ___/ __/   
    / ___ |___/ /_____/ /_/ /  __/ /__/ /_     
   /_/  |_/____/     / .___/\\___/\\___/\\__/     
                    /_/                        }
`);
}
