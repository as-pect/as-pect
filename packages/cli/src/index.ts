import { parse } from "configinator";
import { cliConfig, resolveOptionByName } from "./util/configuration";
import chalk from "chalk";
const pkg = require("../package.json");

export function asp(args: string[], readFileSync: (file: string, basename: string) => string | null): void {
  const configState = parse(args, cliConfig, {
    cwd: process.cwd(),
    readFileSync,
  });

  if (configState.diagnostics.length > 0) {
    for (const diag of configState.diagnostics) {
      console.log(chalk`{red [CLI Error]}: ${diag}`);
    }
    process.exit(1);
  }

  if (resolveOptionByName(configState, "version")) {
    console.log(chalk`{yellow [Version]}: ${pkg.version}`);
    process.exit(0);
  }

  if (!resolveOptionByName(configState, "nologo")) {
    const printAsciiArt = require("./util/asciiArt").printAsciiArt;
    printAsciiArt(pkg.version);
  }

  if (resolveOptionByName(configState, "types")) {
    const types = require("./types").types;
    types();
    process.exit(0);
  }

  if (resolveOptionByName(configState, "portable")) {
    const portable = require("./portable").portable;
    portable();
    process.exit(0);
  }

  process.exit(0);
}

if (typeof require != "undefined" && require.main == module) {
  const fs = require("fs");
  const path = require("path");

  asp(process.argv.slice(2), (file: string, basename: string): string | null => {
    const thePath = path.normalize(path.join(basename, file));
    try {
      return fs.readFileSync(thePath, "utf8");
    } catch (ex) {
      return null;
    }
  });
}