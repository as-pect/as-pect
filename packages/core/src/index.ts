export * from "./reporter/CombinationReporter";
export * from "./reporter/EmptyReporter";
export * from "./reporter/IReporter";
export * from "./reporter/SummaryReporter";
export * from "./reporter/VerboseReporter";
export * from "./test/IWarning";
export * from "./test/TestContext";
export * from "./test/TestNode";
export * from "./util/IAspectExports";
export * from "./util/IPerformanceConfiguration";
export * from "./util/ReflectedValue";
