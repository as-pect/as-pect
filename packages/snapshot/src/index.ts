export * from "./parser";
export * from "./test/Snapshot";
export * from "./test/SnapshotDiff";
export * from "./test/unparse";
