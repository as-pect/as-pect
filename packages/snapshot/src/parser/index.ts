import grammar from "./grammar";
import { Parser, Grammar } from "nearley";

/**
 * Signifies a parsed snapshot.
 */
export type SnapshotData = Map<string, Map<string, Map<string, string>>>;

/**
 * Create a parser manually.
 */
export function createSnapshotParser(): Parser {
  return new Parser(Grammar.fromCompiled(grammar));
}

/**
 * Parse a snapshot string and return a SnapshotData map.
 *
 * @param {string} snapshot- A snapshot stored in stringified format.
 */
export function parseSnapshot(snapshot: string): SnapshotData {
  const parser = createSnapshotParser();
  parser.feed(snapshot);
  const results = parser.results;
  if (results.length !== 1) throw new Error("Invalid snapshot.");
  return results[0] as SnapshotData;
}
