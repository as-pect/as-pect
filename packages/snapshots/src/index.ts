export * from "./Snapshot";
export * from "./SnapshotDiff";
export * from "./SnapshotDiffResult";
