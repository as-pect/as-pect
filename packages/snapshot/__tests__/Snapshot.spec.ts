import { Snapshot, ISnapshotData, ISnapshotStringifyOptions } from "../src";
const example1 = `
exports[\`group 1\`][\`test 1\`][\`snapshot 1\`] = \`
  Array<i32> [1, 2, 3]
\`;
`;
const example2 = `
exports[\`group 1\`][\`test 1\`][\`snapshot 1\`] = \`
  Vec3 {
    x: 1.0,
    y: 2.0,
    z: 3.0,
  }
\`;
`;

const example3: ISnapshotData = {
  group1: {
    test1: {
      snapshot1: "testValue",
      snapshot2: "testValue2",
    },
  },
  group2: {
    test1: {
      snapshot1: "testValue3",
      snapshot2: "testValue4",
    },
  },
  group4: {
    test1: {
      snapshot1: "testValue7",
      snapshot3: "testValue8",
    },
  },
};

const example4: ISnapshotData = {
  group1: {
    test1: {
      snapshot1: "testValue5",
      snapshot2: "testValue6",
    },
  },
  group2: {
    test1: {
      snapshot1: "testValue7",
      snapshot3: "testValue8",
    },
  },
  group3: {
    test1: {
      snapshot1: "testValue7",
      snapshot3: "testValue8",
    },
  },
};

const example5: ISnapshotData = {
  group1: {
    test1: {
      snapshot5: "testValue9",
    },
    test3: {
      snapshot6: "someMatchingValue",
    },
  },
};

const example6: ISnapshotData = {
  group1: {
    test2: {
      snapshot5: "testValue9",
    },
    test3: {
      snapshot6: "someMatchingValue",
    }
  },
};

const options: Partial<ISnapshotStringifyOptions> = {
  addedFormat: str => "added: " + str,
  defaultFormat: str => "default: " + str,
  indent: 2,
  removedFormat: str => "removed: " + str,
};

describe("Snapshots from strings", () => {
  test("Snapshot shape", () => {
    const result1 = Snapshot.fromString(example1);
    expect(result1).toMatchSnapshot("fromstring");

    const result2 = Snapshot.fromString(example2);
    expect(result2).toMatchSnapshot("fromstring");

    const result3 = Snapshot.fromData(example3);
    expect(result3).toMatchSnapshot("fromdata");

    const result4 = Snapshot.fromData(example4);
    expect(result4).toMatchSnapshot("fromdata");

    expect(result1.diff(result2, options)).toMatchSnapshot("diff");
    expect(result3.diff(result4, options)).toMatchSnapshot("diff");

    const result5 = Snapshot.fromData(example5);
    const result6 = Snapshot.fromData(example6);
    expect(result5.diff(result6, options)).toMatchSnapshot("diff");
  });
});
