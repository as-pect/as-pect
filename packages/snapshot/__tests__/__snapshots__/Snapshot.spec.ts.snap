// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots from strings Snapshot shape: diff 1`] = `
Array [
  SnapshotDiff {
    "diff": "  default:   

  removed: -   Array<i32> [1, 2, 3]

  added: +   Vec3 {
    x: 1.0,
    y: 2.0,
    z: 3.0,
  }

",
    "groupName": "group 1",
    "left": "
  Array<i32> [1, 2, 3]
",
    "right": "
  Vec3 {
    x: 1.0,
    y: 2.0,
    z: 3.0,
  }
",
    "snapshotName": "snapshot 1",
    "testName": "test 1",
    "type": 3,
  },
]
`;

exports[`Snapshots from strings Snapshot shape: diff 2`] = `
Array [
  SnapshotDiff {
    "diff": "  removed: - testValue
  added: + testValue5
",
    "groupName": "group1",
    "left": "testValue",
    "right": "testValue5",
    "snapshotName": "snapshot1",
    "testName": "test1",
    "type": 3,
  },
  SnapshotDiff {
    "diff": "  removed: - testValue2
  added: + testValue6
",
    "groupName": "group1",
    "left": "testValue2",
    "right": "testValue6",
    "snapshotName": "snapshot2",
    "testName": "test1",
    "type": 3,
  },
  SnapshotDiff {
    "diff": "  removed: - testValue3
  added: + testValue7
",
    "groupName": "group2",
    "left": "testValue3",
    "right": "testValue7",
    "snapshotName": "snapshot1",
    "testName": "test1",
    "type": 3,
  },
  SnapshotDiff {
    "diff": null,
    "groupName": "group2",
    "left": "testValue4",
    "right": null,
    "snapshotName": "snapshot2",
    "testName": "test1",
    "type": 1,
  },
  SnapshotDiff {
    "diff": null,
    "groupName": "group4",
    "left": "testValue7",
    "right": null,
    "snapshotName": "snapshot1",
    "testName": "test1",
    "type": 1,
  },
  SnapshotDiff {
    "diff": null,
    "groupName": "group4",
    "left": "testValue8",
    "right": null,
    "snapshotName": "snapshot3",
    "testName": "test1",
    "type": 1,
  },
  SnapshotDiff {
    "diff": null,
    "groupName": "group2",
    "left": null,
    "right": "testValue8",
    "snapshotName": "snapshot3",
    "testName": "test1",
    "type": 2,
  },
  SnapshotDiff {
    "diff": null,
    "groupName": "group3",
    "left": null,
    "right": "testValue7",
    "snapshotName": "snapshot1",
    "testName": "test1",
    "type": 2,
  },
  SnapshotDiff {
    "diff": null,
    "groupName": "group3",
    "left": null,
    "right": "testValue8",
    "snapshotName": "snapshot3",
    "testName": "test1",
    "type": 2,
  },
]
`;

exports[`Snapshots from strings Snapshot shape: diff 3`] = `
Array [
  SnapshotDiff {
    "diff": null,
    "groupName": "group1",
    "left": "testValue9",
    "right": null,
    "snapshotName": "snapshot5",
    "testName": "test1",
    "type": 1,
  },
  SnapshotDiff {
    "diff": null,
    "groupName": "group1",
    "left": null,
    "right": "testValue9",
    "snapshotName": "snapshot5",
    "testName": "test2",
    "type": 2,
  },
]
`;

exports[`Snapshots from strings Snapshot shape: fromdata 1`] = `
Snapshot {
  "data": Object {
    "group1": Object {
      "test1": Object {
        "snapshot1": "testValue",
        "snapshot2": "testValue2",
      },
    },
    "group2": Object {
      "test1": Object {
        "snapshot1": "testValue3",
        "snapshot2": "testValue4",
      },
    },
    "group4": Object {
      "test1": Object {
        "snapshot1": "testValue7",
        "snapshot3": "testValue8",
      },
    },
  },
  "stringified": "exports[\`group1\`][\`test1\`][\`snapshot1\`] = \`testValue\`

exports[\`group1\`][\`test1\`][\`snapshot2\`] = \`testValue2\`

exports[\`group2\`][\`test1\`][\`snapshot1\`] = \`testValue3\`

exports[\`group2\`][\`test1\`][\`snapshot2\`] = \`testValue4\`

exports[\`group4\`][\`test1\`][\`snapshot1\`] = \`testValue7\`

exports[\`group4\`][\`test1\`][\`snapshot3\`] = \`testValue8\`

",
}
`;

exports[`Snapshots from strings Snapshot shape: fromdata 2`] = `
Snapshot {
  "data": Object {
    "group1": Object {
      "test1": Object {
        "snapshot1": "testValue5",
        "snapshot2": "testValue6",
      },
    },
    "group2": Object {
      "test1": Object {
        "snapshot1": "testValue7",
        "snapshot3": "testValue8",
      },
    },
    "group3": Object {
      "test1": Object {
        "snapshot1": "testValue7",
        "snapshot3": "testValue8",
      },
    },
  },
  "stringified": "exports[\`group1\`][\`test1\`][\`snapshot1\`] = \`testValue5\`

exports[\`group1\`][\`test1\`][\`snapshot2\`] = \`testValue6\`

exports[\`group2\`][\`test1\`][\`snapshot1\`] = \`testValue7\`

exports[\`group2\`][\`test1\`][\`snapshot3\`] = \`testValue8\`

exports[\`group3\`][\`test1\`][\`snapshot1\`] = \`testValue7\`

exports[\`group3\`][\`test1\`][\`snapshot3\`] = \`testValue8\`

",
}
`;

exports[`Snapshots from strings Snapshot shape: fromstring 1`] = `
Snapshot {
  "data": Object {
    "group 1": Object {
      "test 1": Object {
        "snapshot 1": "
  Array<i32> [1, 2, 3]
",
      },
    },
  },
  "stringified": "
exports[\`group 1\`][\`test 1\`][\`snapshot 1\`] = \`
  Array<i32> [1, 2, 3]
\`;
",
}
`;

exports[`Snapshots from strings Snapshot shape: fromstring 2`] = `
Snapshot {
  "data": Object {
    "group 1": Object {
      "test 1": Object {
        "snapshot 1": "
  Vec3 {
    x: 1.0,
    y: 2.0,
    z: 3.0,
  }
",
      },
    },
  },
  "stringified": "
exports[\`group 1\`][\`test 1\`][\`snapshot 1\`] = \`
  Vec3 {
    x: 1.0,
    y: 2.0,
    z: 3.0,
  }
\`;
",
}
`;
